cowprotocol/cowswap
# feat(tx-bundling): enable bundled approval + presign for Safe App on SWAP

# Summary

Transaction bundling (approval + presign) for Safe App on SWAP

This change makes possible to send a single Safe bundle tx to perform 2 actions: approve + presign


https://github.com/cowprotocol/cowswap/assets/43217/7519a9bb-b5ea-4f58-a5c4-db003f4b3ad0


## ⚠️ IMPORTANT ⚠️ 
* It only works for Safes - other wallets won't be supported for the time being
* It only works when using CoW Swap as a Safe App - WC support is pending on Safe implementation
* It only works for SWAPs - Limit orders coming next

# Testing

1. Load the app inside Safe
2. Pick as sell token one that has 0 approval
* Assuming there's enough liquidity, you should see `approve and swap`
![Screenshot 2023-05-09 at 15 00 35](https://github.com/cowprotocol/cowswap/assets/43217/fa8bd76d-4e40-4a35-b4a2-930a517c329b)

3. Click on it
4. If not on expert mode, the confirmation modal should appear
* (Text hasn't changed yet, will be re-iterated later)
5. Click on it
* You should see the Safe tx pop up, with 2 actions: approve and swap (refer to the video above)
6. Submit the tx
* You should be prompted to sign it in your wallet
* Depending on your Safe setup, the tx will be executed or need more signatures
* Once the tx is executed, the order should change from `pre signing` to `open`

7. Repeat steps `1` and `2` with a Safe connected via WC
* You should see the regular request for approval
8. Repeat steps `1` and `2` with an EOA
* You should see the regular request for approval

# Original

_(please disregard)_

WIP!!!!

Still a huge mess, do not review the code yet.

Right now, this code is creating invalid txs failing to do delegate calls https://dashboard.tenderly.co/public/safe/safe-apps/simulator/560ff74a-606e-4d59-8672-70a7966cc3cf?trace=0.0.0.4.0.1.1

![image](https://user-images.githubusercontent.com/43217/236799915-44cd50c1-4376-4889-9290-0967cae8f070.png)

`msg.sender` is using the Multisend's contract address.

# Testing

Not like it'll work, but this is how you trigger it:

1. Using the Safe, load the app
6. Pick a token pair where the sell token has NO approval
* You should see approve and swap button or something similar
![image](https://user-images.githubusercontent.com/43217/236800676-98c8ffa7-0ceb-465e-baf0-406a1125d7bb.png)
9. I'm also testing with expert mode just so I have one less button to click
10. Click on approve and swap
* It should show you the Safe dialog for the multisend tx like this
![image](https://user-images.githubusercontent.com/43217/236801074-8870271e-fabb-4169-a9ac-37400b73560a.png)
11. Try and simulate the tx
* It SHOULD succeed (but it's where it's currently failing)

*Note:* To make your order history cleaner, I recommend to set tx expiration to 5 min just so the pre-sign txs are expired soon, since they are not executing anyway
